src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,274 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// A stream abstraction over the TDS protocol which works to write the TDS packet
+    /// This stream will take care of wrapping the bytes with the TDS header. 
+    /// </summary>
+    internal class TdsWriteStream : Stream, ITdsWriteStream
+    {
+        private Stream _underlyingStream;
+
+        private byte[] _WriteBuffer;
+
+        // Start at the end of the Tds header
+        internal int WriteBufferOffset { get; private set; } = TdsEnums.HEADER_LEN;
+
+        // This should be set before flushing the buffer.
+        public TdsStreamPacketType PacketHeaderType { get; set; }
+
+        internal byte PacketNumber { get; private set; } = 1; // Packets always start with 1.
+
+        public TdsWriteStream(Stream underLyingStream) : base()
+        {
+            _underlyingStream = underLyingStream;
+            _WriteBuffer = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+        }
+
+        /// <summary>
+        /// Should only be called after the login negotiation is done.
+        /// </summary>
+        /// <param name="bufferSize">new buffer size</param>
+        public void SetPacketSize(int bufferSize) => _WriteBuffer = new byte[bufferSize];
```

Comment:
Food for thought here: is there value in returning the old buffer and renting from ArrayPool?

Reply:
Well, I will err on the side of simplicity here. This method will be called only once during the connection lifecycle, after the new packet size is negotiated with the server. 
So we expect one buffer resize. The default is 8000, and in most cases the customer's dont change the default. 
however there is value in checking if the buffer size has changed from the current size, and then resize the buffer. A micro-optimization, in case the connection string was changed to negotiate a 4096 bytes packet